// AndroidBot/index.js
const wdio = require("webdriverio");
const child_process = require('child_process');
const path = require("path");

/**初始化AndroidBot
 * @param {string} devicesId 设备id
 * @param {number} serverPort AndroidBot服务端口
 * @param {string} package app包名
 * @param {string} activity app活动窗口
 * @param {string} version 安卓版本
 * @param {string} autoName UiAutomator2/UiAutomator1 速度快，不稳定
 * @return {wdio.BrowserObject} 
*/
exports.InitAndroidBot = async (devicesId, serverPort, package, activity, version, autoName = "UiAutomator2")=>{
    await child_process.spawn('cmd',['/c', 'adb -s ' + devicesId + ' shell /data/local/tmp/atx-agent server --stop']);
    child_process.spawn('AndroidBot', ['cmd','/k','appium','--relaxed-security','-p', serverPort, '-bp', serverPort+1, '--log-level','error', process.pid], {'detached':true});
    await this.Sleep(3000);//等待AndroidBot打开通信
    let opts = {
        port: serverPort,
        capabilities: {
            platformName: "Android",
            platformVersion: version,
            deviceName: "AndroidBot",
            udid: devicesId,
            appPackage: package,
            appActivity: activity,
            unicodeKeyboard: "true",
            resetKeyboard: "true",
            noReset: "true",
            newCommandTimeout: "600",
            disableWindowAnimation: "true",
            automationName: autoName
        }
    };
    return await wdio.remote(opts);
}

/**睡眠等待
 * @param {number} millisecond  等待时间,单位毫秒
 * @return {Promise}
*/
exports.Sleep = (millisecond) => {
    return new Promise(resolve => {setTimeout(() => {resolve()}, millisecond)});
  }
